cumulus/pallets/xcmp-queue/src/weights_ext.rs

```diff
@@ -58,36 +59,54 @@ pub trait WeightInfoExt: WeightInfo {
 				.saturating_sub(Self::enqueue_n_bytes_xcmp_message(0))
 		};
 
-		pages_overhead.saturating_add(messages_overhead).saturating_add(bytes_overhead)
+		// If the messages are not added to the beginning of the first page, the page will be
+		// decoded and re-encoded once. Let's account for this.
+		let pos_overhead = {
+			Self::enqueue_empty_xcmp_message_at(first_page_pos)
+				.saturating_sub(Self::enqueue_empty_xcmp_message_at(0))
+		};
+
+		pages_overhead
+			.saturating_add(messages_overhead)
+			.saturating_add(bytes_overhead)
+			.saturating_add(pos_overhead)
+	}
+
+	#[cfg(feature = "std")]
+	fn check_accuracy<MaxMessageLen: bounded_collections::Get<u32>>(err_margin: f64) {
+		assert!(err_margin < 1f64);
+
+		let estimated_weight =
+			Self::uncached_enqueue_xcmp_messages().saturating_add(Self::enqueue_xcmp_messages(
+				get_average_page_pos(MaxMessageLen::get()),
+				&BatchFootprint { msgs_count: 1000, size_in_bytes: 3000, new_pages_count: 0 },
+			));
+		let actual_weight = Self::enqueue_1000_small_xcmp_messages();
+
+		// Check that the ref_time diff is less than err_margin
+		approx::assert_relative_eq!(
+			estimated_weight.ref_time() as f64,
+			actual_weight.ref_time() as f64,
+			max_relative = err_margin
+		);
 	}
 }
 
 impl<T: WeightInfo> WeightInfoExt for T {}
 
-#[cfg(feature = "std")]
-pub fn check_weight_info_ext_accuracy<T: WeightInfoExt>(err_margin: u8) {
-	assert!(err_margin < 100);
-	let err_margin = err_margin as u64;
-
-	let estimated_weight =
-		T::uncached_enqueue_xcmp_messages().saturating_add(T::enqueue_xcmp_messages(1, 1000, 3000));
-	let actual_weight = T::enqueue_1000_small_xcmp_messages();
-
-	// Check that the ref_time diff is less than {err_margin}%
-	let diff_ref_time = estimated_weight.ref_time().abs_diff(actual_weight.ref_time());
-	assert!(diff_ref_time < estimated_weight.ref_time() * err_margin / 100);
-	assert!(diff_ref_time < actual_weight.ref_time() * err_margin / 100);
-
-	// The proof sizes should be the same
-	assert_eq!(estimated_weight.proof_size(), actual_weight.proof_size());
-}
-
 #[cfg(test)]
 mod tests {
 	use super::*;
+	use bounded_collections::ConstU32;
 
 	#[test]
-	fn weight_info_ext_accuracy_is_high() {
-		check_weight_info_ext_accuracy::<()>(5);
+	fn check_weight_info_ext_accuracy() {
+		// The `MaxMessageLen` was manually copied from the `HeapSize` field of the
+		// `asset-hub-westend-runtime`:
+		// https://github.com/paritytech/polkadot-sdk/blob/4173aac8abc7e518d4048306c18d4f2176241206/cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs#L827
+		// It needs to be updated manually in case it changes.
+		// This is a good-enough approximation. To make it 100% accurate, we would need to subtract
+		// the item size. But this is negligible.
+		<() as WeightInfoExt>::check_accuracy::<ConstU32<{ 103 * 1024 }>>(0.05);
```

Comment:
I think we can get this hard-coded `103 * 1024` value from the runtimes somehow. This value can be retrieved at least from two places.
```
impl cumulus_pallet_xcmp_queue::Config for Runtime {
	...
	type MaxPageSize = ConstU32<{ 103 * 1024 }>;
	...
}

impl pallet_message_queue::Config for Runtime {
	...
	type HeapSize = sp_core::ConstU32<{ 103 * 1024 }>;
	...
}
```

Shouldn't we also add some `integrity_test` that `Xcmp::MaxPageSize` is aligned with `MessageQueue::HeapSize` and both are sane?


For DMP we don't have this limit:
```
	type DmpQueue = frame_support::traits::EnqueueWithOrigin<MessageQueue, RelayOrigin>;
```

Reply:
@serban300 can we just make this check as a part of?
```
	#[pallet::hooks]
	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
		fn integrity_test() {
```

Reply:
Oh, very nice suggestion ! Done.

Sorry, I misunderstood the comment the first time.